### ISSUE TITLE
cargo test gives linker errors

### ISSUE THREAD

**Issue by jgilchrist** [tomassedovic/tcod-rs#80]
For example:
```
---- engine::actor::Actor_0 stdout ----
	error: linking with `cc` failed: exit code: 1
note: cc '-Wl,--as-needed' '-m64' '-L' '/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib' '-o' '/tmp/rustdoctest.r1BGY18VAAb0/rust-out' '/tmp/rustdoctest.r1BGY18VAAb0/rust_out.o' '-Wl,--whole-archive' '-lmorestack' '-Wl,--no-whole-archive' '-Wl,--gc-sections' '-pie' '-nodefaultlibs' '/home/jonny/code/rslike/target/librslike-33665a348a9c0d5a.rlib' '/home/jonny/code/rslike/target/deps/libtcod-835af025147faddf.rlib' '/home/jonny/code/rslike/target/deps/libtcod-sys-8caba187ba4adc8b.rlib' '-L' '/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib' '-lstd-4e7c5e5c' '-L' '/home/jonny/code/rslike/target' '-L' '/home/jonny/code/rslike/target/deps' '-L' '/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib' '-L' '/home/jonny/code/rslike/.rust/lib/x86_64-unknown-linux-gnu' '-L' '/home/jonny/code/rslike/lib/x86_64-unknown-linux-gnu' '-Wl,--whole-archive' '-Wl,-Bstatic' '-Wl,--no-whole-archive' '-Wl,-Bdynamic' '-ltcod' '-ltcod' '-ltcod' '-ldl' '-lpthread' '-lrt' '-lgcc_s' '-lpthread' '-lc' '-lm' '-lcompiler-rt'
note: /usr/bin/ld: cannot find -ltcod
/usr/bin/ld: cannot find -ltcod
/usr/bin/ld: cannot find -ltcod
collect2: error: ld returned 1 exit status

error: aborting due to previous error
thread 'engine::actor::Actor_0' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:144
```

For some reason, it's passing the flag `-ltcod` to the linker three times. This doesn't happen when building, only when testing.

**Comment by tomassedovic** [comment]
Hm, strange. I've just updated to latest Rust and I can't reproduce the issue.

Running `cargo test` within `tcod-rs` works as expected. Looking at the error it seems you're seeing it within another app, right? I tried `cargo new` with tcod as a dependency and copied one of the examples. Again `cargo test` just works.

What's your `rustc --version --verbose`? For comparison, mine is:

```
$ rustc --version --verbose
rustc 1.0.0-nightly (8903c21d6 2015-01-15 22:42:58 +0000)
binary: rustc
commit-hash: 8903c21d618fd25dca61d9bb668c5299d21feac9
commit-date: 2015-01-15 22:42:58 +0000
host: x86_64-unknown-linux-gnu
release: 1.0.0-nightly
```

I checked Rust's issues and recent commits. Found a few things that seem related but I can't really tell:

https://github.com/rust-lang/rust/pull/21089

then this, which seems like it's doing something with linking. So this might be a fix or a regression (or completely unrelated):
https://github.com/rust-lang/rust/commit/cbeb77ec7a42b988703fa0b0c857f26d464fdc2f


Other than that, I dunno. Can you try updating to latest Rust master and reproducing? If that fails, it's probably a bug in Rust (third-party code should never crash the compile like this). Any chance you could try get a small reproducible example (ideally a single file + Cargo.toml) and link it here?

I'll try to reproduce & debug it and we can file an issue with Rust.

Oh, also: whenever Rust crashes, try to re-run it with ` RUST_BACKTRACE=1` environment variable set. That should print out a proper backtrace.

And one more thing: I've had a few rust crashes that looked similar with the debugging symbols enabled. Try setting `debug = false` in the dev and test profiles in your `Cargo.toml`. Like here:

https://github.com/tomassedovic/tcod-rs/blob/master/Cargo.toml#L27

It's a long shot, but it worked around the crash for me a few times already.

**Comment by jgilchrist** [comment]
Yeah, I saw this in travis-ci when I pushed a branch on my roguelike.

```
$ rustc --version --verbose
rustc 1.0.0-nightly (8903c21d6 2015-01-15 22:42:58 +0000)
binary: rustc
commit-hash: 8903c21d618fd25dca61d9bb668c5299d21feac9
commit-date: 2015-01-15 22:42:58 +0000
host: x86_64-unknown-linux-gnu
release: 1.0.0-nightly
```

I couldn't reproduce it locally either, `cargo test` ran fine, but then I realised that I have a version of `libtcod` installed locally, so `-ltcod` works even though it shouldn't. Uninstalling the system `libtcod` gives the errors. Turning off debugging symbols didn't do anything.

**Comment by jgilchrist** [comment]
The full error for a single failing test:

```
---- engine::actor::new_0 stdout ----
	error: linking with `cc` failed: exit code: 1
note: cc '-Wl,--as-needed' '-m64' '-L' '/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib' '-o' '/tmp/rustdoctest.sFzeE1E4an78/rust-out' '/tmp/rustdoctest.sFzeE1E4an78/rust_out.o' '-Wl,--whole-archive' '-lmorestack' '-Wl,--no-whole-archive' '-Wl,--gc-sections' '-pie' '-nodefaultlibs' '/home/jonny/code/rslike/target/librslike-33665a348a9c0d5a.rlib' '/home/jonny/code/rslike/target/deps/libtcod-835af025147faddf.rlib' '/home/jonny/code/rslike/target/deps/libtcod-sys-8caba187ba4adc8b.rlib' '-L' '/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib' '-lstd-4e7c5e5c' '-L' '/home/jonny/code/rslike/target' '-L' '/home/jonny/code/rslike/target/deps' '-L' '/usr/local/lib/rustlib/x86_64-unknown-linux-gnu/lib' '-L' '/home/jonny/code/rslike/.rust/lib/x86_64-unknown-linux-gnu' '-L' '/home/jonny/code/rslike/lib/x86_64-unknown-linux-gnu' '-Wl,--whole-archive' '-Wl,-Bstatic' '-Wl,--no-whole-archive' '-Wl,-Bdynamic' '-ltcod' '-ltcod' '-ltcod' '-ldl' '-lpthread' '-lrt' '-lgcc_s' '-lpthread' '-lc' '-lm' '-lcompiler-rt'
note: /usr/bin/ld: cannot find -ltcod
/usr/bin/ld: cannot find -ltcod
/usr/bin/ld: cannot find -ltcod
collect2: error: ld returned 1 exit status

error: aborting due to previous error
thread 'engine::actor::new_0' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:144

stack backtrace:
   1:     0x7f0d8802c000 - sys::backtrace::write::h7e8ba1fc274cc36epRt
   2:     0x7f0d8804d9e0 - failure::on_fail::h823798930ebe140a64z
   3:     0x7f0d87fbc7b0 - rt::unwind::begin_unwind_inner::ha4a8df6010104da6ZJz
   4:     0x7f0d83046cd0 - rt::unwind::begin_unwind::h15231093072926810949
   5:     0x7f0d830474c0 - diagnostic::Handler::fatal::hf2fba868154a7d81Q0F
   6:     0x7f0d83047730 - diagnostic::Handler::abort_if_errors::h49ed039430cd670by2F
   7:     0x7f0d871282d0 - back::link::link_natively::hf78742a8a3f2054bCXa
   8:     0x7f0d87120820 - back::link::link_binary::hd00660e13e96680bSua
   9:     0x7f0d87611270 - driver::phase_6_link_output::hcf0e359de93d6713rPa
  10:     0x7f0d875efd70 - driver::compile_input::h8ec974b2bd4afe59Aba
  11:     0x7f0d8866c530 - test::runtest::h51d251a4024cad3bj2t
  12:     0x7f0d88677320 - thunk::F.Invoke<A, R>::invoke::h1954005010071349229
  13:     0x7f0d87930050 - thunk::F.Invoke<A, R>::invoke::h12312287147979335838
  14:     0x7f0d87926a50 - thunk::F.Invoke<A, R>::invoke::h10273972757065453811
  15:     0x7f0d87927200 - rt::unwind::try::try_fn::h6611577649401309552
  16:     0x7f0d880b4b90 - rust_try_inner
  17:     0x7f0d880b4b80 - rust_try
  18:     0x7f0d879274b0 - thunk::F.Invoke<A, R>::invoke::h12348411253666505759
  19:     0x7f0d8803b680 - sys::thread::thread_start::h2ac456e96ea03cedcJw
  20:     0x7f0d82862250 - start_thread
  21:     0x7f0d87c6d219 - clone
  22:                0x0 - <unknown>
```

**Comment by jgilchrist** [comment]
rust-lang/rust#21246 looks like the issue. Potentially a bug in Cargo.

**Comment by tomassedovic** [comment]
Ugh :-/

In case you missed, I've noticed this workaround they did in Servo:

https://github.com/servo/servo/pull/4643/files

**Comment by jgilchrist** [comment]
I saw, but we can't really ask people who use this library to disable their doctests. There's been no activity on the issue in Rust, so I'm not really sure the deal is.

**Comment by tomassedovic** [comment]
Yeah, good point.

**Comment by jgilchrist** [comment]
The problem seems to be that the tcod-sys output directory is not passed to rustdoc:

The call to rustc:
```
rustc src/lib.rs --crate-name rslike --crate-type lib -g -C metadata=33665a348a9c0d5a -C extra-filename=-33665a348a9c0d5a --out-dir /home/jonny/code/rslike/target --emit=dep-info,link 
-L dependency=/home/jonny/code/rslike/target 
-L dependency=/home/jonny/code/rslike/target/deps 
--extern tcod=/home/jonny/code/rslike/target/deps/libtcod-ced67c17ab23ef28.rlib 
-L /home/jonny/code/rslike/target/build/tcod-sys-f8f3021d7cdb0f8e/out
```

The call to rustdoc:
```
rustdoc --test /home/jonny/code/rslike/src/lib.rs --crate-name rslike 
-L /home/jonny/code/rslike/target -L /home/jonny/code/rslike/target/deps 
--extern bitflags=/home/jonny/code/rslike/target/deps/libbitflags-518ea12e21428edd.rlib 
--extern tcod-sys=/home/jonny/code/rslike/target/deps/libtcod-sys-f8f3021d7cdb0f8e.rlib 
--extern tcod=/home/jonny/code/rslike/target/deps/libtcod-ced67c17ab23ef28.rlib 
--extern rslike=/home/jonny/code/rslike/target/librslike-33665a348a9c0d5a.rlib
```

**Comment by jgilchrist** [comment]
Fixed with rust-lang/cargo#1530
